rkt/containers.go

```diff
@@ -0,0 +1,178 @@
+// Copyright 2014 CoreOS, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"fmt"
+	"io/ioutil"
+	"os"
+	"path/filepath"
+
+	"github.com/coreos/rocket/pkg/lock"
+)
+
+type containerHandle struct {
+	uuid       string
+	lock       *lock.DirLock
+	isExited   bool
+	isGarbage  bool
+	isDeleting bool
+}
+
+type includeMask byte
+
+const (
+	includeContainersDir includeMask = 1 << iota
+	includeGarbageDir
+)
+
+// withContainerHandles iterates over the included containers calling function f for every container.
+func withContainerHandles(include includeMask, f func(*containerHandle) error) error {
+	ls, err := getContainers(include)
+	if err != nil {
+		return fmt.Errorf("failed to get containers: %v", err)
+	}
+
+	for _, uuid := range ls {
+		ch, err := getContainerHandle(uuid)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "Error getting container handle %q, ignoring\n", uuid)
+			continue
+		}
+		// TODO(vc): how to respond to f() errors, what's idiomatic for this pattern?
+		f(ch)
+		freeContainerHandle(ch)
+	}
+	return nil
+}
+
+// getContainerHandle returns a ContainerHandle struct representing the given container.
+// The in the returned handle is always left in an open but unlocked state.
+// The handle must be freed using freeContainerHandle.
+func getContainerHandle(uuid string) (*containerHandle, error) {
+	s := &containerHandle{}
+
+	l, err := lock.NewLock(filepath.Join(containersDir(), uuid))
+	if err == lock.ErrNotExist {
+		l, err = lock.NewLock(filepath.Join(garbageDir(), uuid))
+		s.isGarbage = true
+		s.isExited = true
+	}
+
+	if err != nil {
+		return nil, fmt.Errorf("error opening container %q: %v", uuid, err)
+	}
+
+	if err = l.TrySharedLock(); err != nil {
+		if err != lock.ErrLocked {
+			l.Close()
+			return nil, fmt.Errorf("unexpected lock error: %v", err)
+		}
+		if s.isGarbage {
+			s.isDeleting = true
+		}
+	} else {
+		l.Unlock()
+		s.isExited = true
+	}
+
+	s.uuid = uuid
+	s.lock = l
+
+	return s, nil
+}
+
+// freeContainerHandle closes the associated lock of a container handle opened via getContainerHandle.
```

Comment:
if you promote the lock as mentioned above, this function is obviated imho, users of a containerhandle can just call Close()